### PR TITLE
feat(transformer): support react fast refresh

### LINKED ISSUE #3943: transformer: React fast refresh transform
Reference: 

* https://www.npmjs.com/package/@vitejs/plugin-react-swc
* https://github.com/facebook/react/blob/5b89d353e27a7ef07b40929f7343ebcc43bdaa58/packages/react-refresh/src/ReactFreshBabelPlugin.js
* https://github.com/pmmmwh/react-refresh-webpack-plugin
* https://mariosfakiolas.com/blog/what-the-heck-is-react-fast-refresh/

### PR DESCRIPTION
**Pull request by Dunqing**
close: #3943

## Further improvements

There is a double visit here. We need to collect all react hooks calling in `Function` and `ArrowFunctionExpression`. If we want to remove this implementation, we may wait for #4188.

https://github.com/oxc-project/oxc/blob/d797e595d286c613848b773c256bd43124ad1981/crates/oxc_transformer/src/react/refresh.rs#L744-L947

## Tests

All tests copy from https://github.com/facebook/react/blob/main/packages/react-refresh/src/__tests__/ReactFresh-test.js

There are still 4 tests that have not been passed

**1. refresh/can-handle-implicit-arrow-returns/input.jsx**

Related to #4767. transform correct, just output doesn't match the expected output

**2. refresh/registers-identifiers-used-in-jsx-at-definition-site/input.jsx**
**3. refresh/registers-identifiers-used-in-react-create-element-at-definition-site/input.jsx**

Blocked by #4746 

**4. refresh/supports-typescript-namespace-syntax/input.tsx** 

oxc transforms ts to js first, so probably we can ignore this case. If we really want to pass this test, we also need to turn off `TypeScript` plugin.

### Options:

1. Support transform `refresh_reg` and `refresh_sig` options to `MemberExpression`. Currently `import.meta.xxxx` still is an `Identifier`
2. Support `emit_full_signatures` option

### Other
NAPI, testing in `monitor-oxc`, etc..

### CODE REVIEW

**Comment by Dunqing** [comment]
* **#4587** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4587?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Dunqing and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by overlookmotel** [comment]
Is this ready for review? Or blocked on #4746 and #4767?

**Comment by Boshen** [comment]
> Is this ready for review? Or blocked on #4746 and #4767?

Ready for review. Can be merged before fixing these.

**Review by overlookmotel — COMMENTED** [review]
As discussed, I haven't gone fully into this, so I haven't sought to understand all the logic of the plugin, or this implementation. But I get at least a vague sense that *maybe* there's some duplicated logic in this implementation, especially in `transform_statements` / `transform_statements_on_exit`.

I may be wrong about that, but could you please check?

Just to be clear, I don't object to merging this - we could always merge and then iterate. But you've asked me to review, so I'm just giving my thoughts.

Wider point: I think we need a better naming convention to make it easier to recognise which functions are "entry points" from the visitor, and which are functions which are only called *by* those entry points. I think if we call the entry points `transform_*` and `transform_*_on_exit`, we should use different names for the internal methods. When everything is called `transform_*`, it makes it quite hard to follow the logic.

File: crates/oxc_transformer/src/react/refresh.rs
Comment (Dunqing) [reply crates/oxc_transformer/src/react/refresh.rs:408]: We have to keep it for now, because we may need to insert new statements under the current statement. Related to #4767.

Both `ExportNamedDeclaration` and `ExportDefaultDeclaration` may confuse you. But we also need to transform the `FunctionDeclaration` inside the `BlockStatement`

For example: 
```ts
function Foo() {
  function name() {}
}
while (true)
  function name() {}
}
```

File: crates/oxc_transformer/src/react/refresh.rs
Comment (Dunqing) [reply crates/oxc_transformer/src/react/refresh.rs:660]: Yes, that's right, but in my opinion, using `Ancestor::VariableDeclaratorInit` is not intuitive. I know what it is,  but others may be confused.

**Comment by Dunqing** [comment]
> Wider point: I think we need a better naming convention to make it easier to recognise which functions are "entry points" from the visitor, and which are functions which are only called by those entry points. I think if we call the entry points transform_* and transform_*_on_exit, we should use different names for the internal methods. When everything is called transform_*, it makes it quite hard to follow the logic.

Yes! I agree, I'll try to rename these methods

File: crates/oxc_transformer/src/react/refresh.rs
Comment (overlookmotel) [reply crates/oxc_transformer/src/react/refresh.rs:749]: I agree `matches!(ctx.parent(), Ancestor::VariableDeclaratorInit(_))` is pretty ugly. Shall we add `is_variable_declarator_init()` method? (and same for all `Ancestor` variants, though it'd be a lot of methods to generate)

I can see that `Ancestor`'s pattern of including both the parent type *and* the field in its variants is unusual compared to Oxc's other APIs. But it's central to how `Traverse` works (required for safety). So basically... anyone who wants to work on transformer will have to get used to it! In my view, it preferable if we make the transforms we write exemplary of best API usage, so others can copy them.

File: crates/oxc_transformer/src/react/refresh.rs
Comment (overlookmotel) [reply crates/oxc_transformer/src/react/refresh.rs]: Ah ha I see. Thanks for explaining.

File: crates/oxc_transformer/src/react/refresh.rs
Comment (Dunqing) [reply crates/oxc_transformer/src/react/refresh.rs:749]: > I agree `matches!(ctx.parent(), Ancestor::VariableDeclaratorInit(_))` is pretty ugly. Shall we add `is_variable_declarator_init()` method? (and same for all `Ancestor` variants, though it'd be a lot of methods to generate)
> 
Yes, we should add it later.

> I can see that `Ancestor`'s pattern of including both the parent type _and_ the field in its variants is unusual compared to Oxc's other APIs. But it's central to how `Traverse` works (required for safety). So basically... anyone who wants to work on transformer will have to get used to it! In my view, it preferable if we make the transforms we write exemplary of best API usage, so others can copy them.

I agree, it sounds reasonable.

**Comment by Dunqing** [comment]
@overlookmotel Please take a look again, once this is merged, I will start to refactor it according to #4881 and try to resolve the remaining tests

**Review by overlookmotel — APPROVED** [review]
I can't review this properly without delving fully into how the transform works.

I agree that we should merge it now. If you're going to refactor it anyway in a follow-on, it'll be easier for me to do a "2nd pass" review after that.

Currently there are merge conflicts. If you're able to resolve those, please go ahead and merge.

**Pull request merged**
